cmd/sync-rover-groups/yamlgroupcollector.go

```diff
@@ -0,0 +1,206 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	templatev1 "github.com/openshift/api/template/v1"
+	userv1 "github.com/openshift/api/user/v1"
+)
+
+type yamlGroupCollector struct {
+	decoder          runtime.Decoder
+	validateSubjects bool
+}
+
+const yamlSeparator = "\n---"
+
+func (c *yamlGroupCollector) collect(dir string) (sets.String, error) {
+	groups := sets.NewString()
+	abs, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine absolute path for dir %s: %w", dir, err)
+	}
+	if err := filepath.Walk(abs,
```

Comment:
This whole thing below is IMO a bit hard to read, too long and too many abstraction levels (indents and nested cycles). Some proposals inline about bits that could perhaps be extracted into separate methods.